### PR TITLE
New ConvertHFIRSCDtoMDE algorithm for WAND and DEMAND

### PR DESCRIPTION
**Pull request by rosswhitfield**
**Description of work.**

This algorithm has the same functionality as [ConvertWANDSCDtoQ](https://docs.mantidproject.org/nightly/algorithms/ConvertWANDSCDtoQ-v1.html) except that it convert to a MDEventWorkspace instead of MDHistoWorkspace, and only to Q_sample with no normalization.

The advantage of using an MDEventWorkpsace is it should provided more acturate results from FindPeaksMD and will also enable the use of Peak integration algorithm in Mantid that only work for MDEventWorksapces, like [IntegratePeaksMD](https://docs.mantidproject.org/nightly/algorithms/IntegratePeaksMD-v2.html)

**To test:**

This should work with both WAND and DEMAND data. View the workspaces in sliceviewer and check that they look correct.

WAND example, you can also try the usage example for another WAND test data set.
```python
data = LoadWANDSCD(IPTS=22745, RunNumbers='147131-148931', Grouping='4x4')
mde=ConvertHFIRSCDtoMDE(data, Wavelength=1.488)
```

DEMAND example, you should get the same results as in #27536
```python
data2 = Load("/HFIR/HB3A/IPTS-24655/shared/autoreduce/HB3A_exp0724_scan0183.nxs")
mde2=ConvertHFIRSCDtoMDE(data2, Wavelength=1.008)
```

*There are no issues.*

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Testing/SystemTests/tests/analysis/ConvertHFIRSCDtoMDETest.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+#     NScD Oak Ridge National Laboratory, European Spallation Source
+#     & Institut Laue - Langevin
+# SPDX - License - Identifier: GPL - 3.0 +
+import systemtesting
+import numpy as np
+from mantid.simpleapi import *
+
+
+class ConvertHFIRSCDtoMDETest(systemtesting.MantidSystemTest):
+    def requiredMemoryMB(self):
```
Comment (peterfpeterson) [inline Testing/SystemTests/tests/analysis/ConvertHFIRSCDtoMDETest.py]: Thanks for letting people know

File: Framework/MDAlgorithms/src/ConvertHFIRSCDtoMDE.cpp [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+    size_t number_of_runs = inputWS->getDimension(2)->getNBins();
+    std::vector<std::string> logs;
+    if (instrument == "HB3A")
+      logs = {"omega", "chi", "phi", "monitor", "time"};
+    else
+      logs = {"duration", "monitor_count", "s1"};
+    for (auto log : logs) {
+      if (run.hasProperty(log)) {
+        if (static_cast<size_t>(run.getLogData(log)->size()) != number_of_runs)
+          inputWSmsg << "Log " << log << " has incorrect lenght, ";
```
Comment (peterfpeterson) [inline Framework/MDAlgorithms/src/ConvertHFIRSCDtoMDE.cpp]:
```suggestion
          inputWSmsg << "Log " << log << " has incorrect length, ";
```

File: Framework/MDAlgorithms/src/ConvertHFIRSCDtoMDE.cpp [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+                      "OutputWorkspace", "", Direction::Output),
+                  "An output workspace.");
+}
+
+//----------------------------------------------------------------------------------------------
+/** Execute the algorithm.
+ */
+void ConvertHFIRSCDtoMDE::exec() {
+  double wavelength = this->getProperty("Wavelength");
+  if (wavelength == Mantid::EMPTY_DBL()) {
```
Comment (peterfpeterson) [inline Framework/MDAlgorithms/src/ConvertHFIRSCDtoMDE.cpp]: Isn't this taken care of by the bounded validator in `init`?

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
Two minor changes. Otherwise this is ready to go. I did not verify the equations for the goniometer rotations, but they peaks were in the correct place as I scanned Qy in slice-viewer.

File: Framework/MDAlgorithms/src/ConvertHFIRSCDtoMDE.cpp (hunk 3 above)
Comment (rosswhitfield) [reply Framework/MDAlgorithms/src/ConvertHFIRSCDtoMDE.cpp]: Yes, that was there from before I added it to the validation so I can remove it.

**Review by peterfpeterson — APPROVED** [review]

**Pull request merged**